AddressDataType/AddressFormatter.cs

```diff
@@ -1,71 +1,76 @@
+    /// <param name="addressFormat">
+    /// A format string for the address. Parameters are, in order:
+    /// 0. The name.
+    /// 1. The department.
+    /// 2. The company or institution.
+    /// 3. The first line of the street address.
+    /// 4. The second line of the street address.
+    /// 5. The city line (<see cref="_cityLineFormat"/>).
+    /// 6. The country.
+    /// </param>
+    /// <param name="cityLineFormat">
+    /// A format string for the city line of the address. Parameters are, in order:
+    /// 0. The city.
+    /// 1. The province or state.
+    /// 2. The postal code.
+    /// </param>
+    /// <param name="uppercase">If <see langword="true"/>, the address is changed to uppercase after formatting.</param>
+    public AddressFormatter(
+        string addressFormat = @"{0}
 {1}
 {2}
 {3}
 {4}
 {5}
 {6}",
-            string cityLineFormat = "{0} {1} {2}",
-            bool uppercase = true)
-        {
-            _addressFormat = addressFormat;
-            _cityLineFormat = cityLineFormat;
-            _uppercase = uppercase;
-        }
+        string cityLineFormat = "{0} {1} {2}",
+        bool uppercase = true)
+    {
+        _addressFormat = addressFormat;
+        _cityLineFormat = cityLineFormat;
+        _uppercase = uppercase;
+    }
+
+    /// <summary>
+    /// Formats the address with the format strings provided.
+    /// </summary>
+    public string Format(Address address)
+    {
+        if (address is null) return "-";
 
-        /// <summary>
-        /// Formats the address with the format strings provided
-        /// </summary>
-        /// <param name="address"></param>
-        /// <returns></returns>
-        public string Format(Address address)
-        {
-            if (address is null) return "-";
-            string rawFormatted = String.Format(
-                _addressFormat,
-                address.Name,
-                address.Department,
-                address.Company,
-                address.StreetAddress1,
-                address.StreetAddress2,
-                String.Format(_cityLineFormat, address.City, address.Province, address.PostalCode),
-                address.Region);
-            string withoutEmptyLines = String.Join("", rawFormatted.Split(Environment.NewLine, StringSplitOptions.RemoveEmptyEntries));
-            return _uppercase ? withoutEmptyLines.ToUpper() : withoutEmptyLines;
-        }
+        var rawFormatted = string.Format(
+            CultureInfo.InvariantCulture,
+            _addressFormat,
+            address.Name,
+            address.Department,
+            address.Company,
+            address.StreetAddress1,
+            address.StreetAddress2,
+            string.Format(CultureInfo.InvariantCulture, _cityLineFormat, address.City, address.Province, address.PostalCode),
+            address.Region);
+        var withoutEmptyLines = Regex
+            .Replace(rawFormatted, @"\n\n+", "\n", RegexOptions.None, TimeSpan.FromSeconds(1))
```

Comment:
I don't think this data is from the file system so you wouldn't have `\r` in the first place. But all right then, this will retain the first line ending:
```csharp
        var withoutEmptyLines = Regex
            .Replace(rawFormatted, @"(?<first>\r?\n)[\r\n]+", "${first}", RegexOptions.ExplicitCapture, TimeSpan.FromSeconds(1))
            .Trim('\r', '\n');
```

Reply:
Very nice!

I think the data may come from the FS because there's an inlined default higher up in the same class. That would contain CRLF if we have to keep the OS-specific line endings.